### DISCUSSION TITLE
send() failed (32: Broken pipe) when performing HTTP POST of a large (>1MB) file

### DISCUSSION THREAD

**Discussion by Matthias-vdE** [ZoeyVid/NPMplus#1123] [Q&A] [answered]
I have an instance of Immich behind NPMplus and while uploading pictures and videos directly via IP address works fine, I'm having issues when going via the reverse proxy. I see this in the log:

```
2024-09-27T11:33:13.859502771Z 2024/09/27 13:33:13 [error] 32855#32855: *10529 send() failed (32: Broken pipe), client: MY_IPADDRESS, server: immich.mydomain.org, request: "POST /api/assets HTTP/1.1", host: "immich.mydomain.org"

2024-09-27T11:33:13.859545312Z 2024/09/27 13:33:13 [error] 32855#32855: *10529 [lua] crowdsec.lua:578: AppSecCheck(): Fallback because of err: broken pipe, client: MY_IPADDRESS, server: immich.mydomain.org, request: "POST /api/assets HTTP/1.1", host: "immich.mydomain.org"

2024-09-27T11:33:13.859553917Z 2024/09/27 13:33:13 [error] 32855#32855: *10529 [lua] crowdsec.lua:651: Allow(): AppSec check: broken pipe, client: MY_IPADDRESS, server: immich.mydomain.org, request: "POST /api/assets HTTP/1.1", host: "immich.mydomain.org"

2024-09-27T11:33:13.859564759Z 2024/09/27 13:33:13 [alert] 32855#32855: *10529 [lua] crowdsec.lua:718: Allow(): [Crowdsec] denied 'MY_IPADDRESS' with 'ban' (by appsec), client: MY_IPADDRESS, server: immich.mydomain.org, request: "POST /api/assets HTTP/1.1", host: "immich.mydomain.org"
```

The first line gives a send() failed (32: Broken pipe) error, which in turn causes Crowdsec to drop the connection.

A broken pipe issue should generally be fixed by setting a larger client_max_body_size and proxy_timeout parameters which I did. Full config here:

![image](https://github.com/user-attachments/assets/db087dc8-60f1-403f-831f-cf56a7bc8256)

(nothing in Custom locations. TLS only forces HTTPS with a certbot certificate)

![image](https://github.com/user-attachments/assets/3681979b-8ac7-4ba0-8b05-89e3edc356f2)

(Not sure exactly if this an NPMplus or Immich specific issue, but as mentioned, not going via the reverse proxy doesn't give issues. But come to think of it, I'm also running an instance of Paperless which is more than happy accepting 100MB+ PDF files without issue with a similar config...)

**Comment by Zoey2936** [comment]
See https://github.com/ZoeyVid/NPMplus/discussions/873

**Reply by Matthias-vdE** [reply]
Been searching high and low for an answer but completely seem to have missed that post.

Removing these from the crowdsec.conf fixed it...

```
APPSEC_URL=http://crowdsec:7422
APPSEC_FAILURE_ACTION=deny
```

Is there a way to disable Appsec for a single host? Modsecurity is not enabled for this host either because it gave some issues.

**Reply by Zoey2936** [reply]
sorry, but that is not possible, but you could also change deny to allow, but that would mean that allow non cheked request would be allowed.

**Reply by Matthias-vdE** [reply]
Ok, thanks. Just for my understanding: Does this completely disable Crowdsec and/or Modsecurity, or only the Appsec component?

**Reply by Zoey2936** [reply]
Only appsec

**Reply by Zoey2936** [reply]
The appsec issue should be fixed in the next release see: https://github.com/ZoeyVid/NPMplus/discussions/873#discussioncomment-10991230

**Reply by Matthias-vdE** [reply]
I just did a test and confirm it seems to work. I was able to upload a 200MB file without issues. Will test with an even larger file later.

**Reply by Matthias-vdE** [reply]
1GB+ files still gives the following message:

`2024-10-22T10:07:03.063676335Z 2024/10/22 12:06:46 [error] 3217#3217: *2122 Request body limit is marked to reject the request, client: 94.X.X.X, server: immich.x.x, request: "POST /api/assets HTTP/1.1", host: "immich.x.x"`

client_max_body_size is set to 0.

**Reply by Zoey2936** [reply]
Modsec is disabled?

**Reply by Matthias-vdE** [reply]
Yeah OK that fixed it again lol. I wrongfully assumed the Appsec fix automagically fixed Modsec as well.

**Reply by Zoey2936** [reply]
No, but you can try to edit the modsec limits in its config file

**Reply by Matthias-vdE** [reply]
Unfortunately the issue returned out of the blue with the same "broken pipe" error message. I have the new timings in my config file, as I completely removed and recreated the Crowdsec integration (and confirm it worked again 2 days ago). Removing

```
APPSEC_URL=http://crowdsec:7422
APPSEC_FAILURE_ACTION=deny
```

From the Crowdsec.conf file fixes it again.

I'll leave appsec and crowdsec disabled for the time being, which fixes this particular issue with Immich.

**Comment by yurividal** [comment]
Having the same issue with immich.
Was this ever fixed?
Any workarrounds?

**Reply by Zoey2936** [reply]
if you read this discussion you will find workarrounds

**Reply by yurividal** [reply]
I did, but it seems that I need to disable crowdsec? Is it not possible to have immich + crowdsec working then?

**Reply by Zoey2936** [reply]
Not crowdsec, but appsec. Can you send me your crowdsec cof (without the api key), please?

**Reply by yurividal** [reply]
Ok, I solved it by disabling modsecurity on the GUI of that specific proxy host in NPMPlus.
Works now.
I left modsecurity on for the other services that don't require any file uploads

**Comment by Zoey2936** [comment, marked as answer]
The Issue seems to still exist, please continue any discussions in https://github.com/ZoeyVid/NPMplus/discussions/1241, if you have any further useful information you can also leave it here: https://github.com/crowdsecurity/lua-cs-bouncer/issues/77
